### DISCUSSION TITLE
I'd like to put this working PowerShell script in the Scripts folder

### DISCUSSION THREAD

**Discussion by JJKylee** [staxrip/staxrip#603] [Q&A] [answered]
I've composed this one-liner that produces a text file containing the _frame numbers of I-frames_ in each line given an input video file.

> Input video file: `INPUT.mkv`
> Output file: `INPUT.mkv I-frame list.txt` (generated automatically based on the input file name)
> Prerequisite: `ffprobe.exe` should be in `PATH`.

```PowerShell
$filename = "INPUT.mkv"; ffprobe -loglevel error -select_streams v:0 -show_entries frame=pict_type -of flat $filename | Select-String "I" -CaseSensitive | ForEach-Object { Write-Output ($_.LineNumber-1) } | Out-File -FilePath $filename" I-frame list.txt" -Encoding ascii
```

This is a PowerShell version of [these CMD and Bash scripts](https://forum.doom9.org/showthread.php?p=1938227#post1938227).

I wish to put this script in StaxRip so that users can get an I-frame list text file in the same folder as the source file after a source file is loaded.

But I have two hurdles.

1. I'd like to have the script get the source file name automatically. How can I apply the `%source_file%` macro in the script?
2. How can I deal with the situation where the user does not have `ffprobe.exe` in `PATH`?

Any help is very welcome. But in particular, I'd like to hear from @stax76 and @Dendraspis. 😄

---

### Update

After posting the script, I found that there was an issue in the `ffprobe` options part, so I fixed it.
Removed `-show_frames` and changed the output format(`-of`) from `csv` to `flat`.
And added the `-CaseSensitive` switch accordingly for the PowerShell cmdlet `Select-String`.
The script is updated.

For a reference, this script took me about 2 mins 28 seconds to yield the I-frame list of the '[Chameleon](https://we.tl/t-Sr7tLrHR8L)' HEVC-m2ts Dolby Vision trailer.

[Chameleon.m2ts I-frame list.txt](https://github.com/staxrip/staxrip/files/6169783/Chameleon.m2ts.I-frame.list.txt)

**Comment by ghost** [comment]
Maybe as part of the FFMPEG package, FFProbe should be included in the release... I know that would mean also another file to move when switching from portable AviSynth to system AviSynth. But it could be good to have it...

**Reply by JJKylee** [reply]
Since `ffprobe.exe` is usually not used in reading/piping `avs` scripts, I don't think we need to move it around depending on which AviSynth version is used. But I'm wondering if there's any use case where `ffprobe.exe` itself needs to read an `avs` script. 🙄

**Reply by ghost** [reply]
Oh yes you're right, and ffprobe does not need ffmpeg, so this is no problem !

**Comment by JJKylee** [comment]
## Critical issue with using `ffprobe`

After looking further into this matter, I've found a _critical_ issue in this approach: `ffprobe` takes _too much time_ in yielding the I-frame frame number list. It generally takes longer than the playback time of the source video. As I mentioned in the original post, in the case of '[Chameleon](https://we.tl/t-Sr7tLrHR8L)' it takes 2 mins 28 seconds while the playback time is only 1 min 3 seconds. 😓

This is because `ffprobe` runs through _all_ the frames in order to get the indexes of frames. That is, `ffprobe` serves as an _indexer_ in this case.

We cannot skip non-I frames though because frame indexes(=frame numbers) are not an _hard-wired property_ of a frame packet. Therefore in order to get the frame index numbers, we have to run through _all_ the frames in a video stream. (And that's why an indexer is required for frame-based video processing in AviSynth+ or VapourSynth.)

At this point, the merit of using `ffprobe` disappears.

## Alternative approach

However, we already have _good_ indexers. `FFVideoSource`, `LSMASHVideoSource`, `DGIndex`, `DGIndexNV`, etc. are examples of those indexers specialized in specific video stream types. We don't have to reinvent the wheel from scratch.

AFAIK, among those indexers, only `DGIndex` and `DGIndexNV` provide _text-based_ index files: `d2v` and `dgi` respectively. Other indexers either encrypt the index file (`ffindex` for `FFVideoSource`) or hide it inside the memory (`LSMASHVideoSource`).

Since these text index files (`d2v` and `dgi`) contain `pict_type` information for each indexed frame, it's easy for us to parse the index file to extract the I-frame indexes. `DGIndex GUI` and `DGIndexNV GUI` make use of this `pict_type` info in their preview windows, but they don't provide a feature to extract I-frame index numbers. And of course StaxRip is ignorant of this `pict_type` info provided by the index files since it only makes use of the decoded frames served by the source filters (`MPEG2Source` and `DGSource`, respectively.) So this 'I-frame index number extraction job' can be a meaningful task.

However since `DGIndex` is limited only to MPEG-2 video streams, I think parsing `dgi` is more meaningful although its use is limited to licensed users only.

So long story short, I'll be using `dgi` (generated by `DGIndexNV`), not `ffprobe` to extract I-frame indexes.

## The PowerShell script

Depending on the video stream, the `pict_type` is placed at different locations according to the `DGI File Structure` manual in `DGIndexNVManual.html` (included in the `DGDecNV` package.) So we need to specify where it is located, and what is the symbol representing the I-frame.

### AVC

Each line of frame data (one line per frame) has this format:

> `n:FRM w x y`

A typical line is...

> `0:FRM 7 0 0 5`

Here, the second field separated by a space represents the `pict_type` and `7` is the symbol for I-frame.
And since only frame data lines have this format, we can identify them with the regular expression `^[0-9]+:`.
As explained, per each captured line, we just need to get the index number for lines where the second field separated by a space matches `7`.

So the PowerShell script becomes like this:

Assuming,

> Input file: `INPUT.dgi`
> Output file: `INPUT.dgi I-frame list.txt`

```PowerShell
$filename = "INPUT.dgi"
$iframelist = $filename + " I-frame list.txt"
Select-String -Path $filename "^[0-9]+:" | ForEach-Object { if ($_.Line.split(" ")[1] -eq '7') { Write-Output $_.Line.split(":")[0] } } | Out-File -FilePath $iframelist -Encoding ascii
```

### HEVC

The second last field separated by a space represents the `pict_type` and `2` is the symbol for I-frame.

> `n:IDR/CRA/BLA n w x`

or

> `n:TRAIL/RASL/RADL w x`

```PowerShell
$filename = "INPUT.dgi"
$iframelist = $filename + " I-frame list.txt"
Select-String -Path $filename "^[0-9]+:" | ForEach-Object { if($_.Line.split(" ")[-2] -eq '2') { Write-Output $_.Line.split(":")[0] } } | Out-File -FilePath $iframelist -Encoding ascii
```

### MPG

Very similar to AVC except the I-frame symbol is `I`.

```PowerShell
$filename = "INPUT.dgi"
$iframelist = $filename + " I-frame list.txt"
Select-String -Path $filename "^[0-9]+:" | ForEach-Object { if($_.Line.split(" ")[1] -eq 'I') { Write-Output $_.Line.split(":")[0] } } | Out-File -FilePath $iframelist -Encoding ascii
```

### VC1

Very similar to AVC except that the I-frame symbol is `3`.

```PowerShell
$filename = "INPUT.dgi"
$iframelist = $filename + " I-frame list.txt"
Select-String -Path $filename "^[0-9]+:" | ForEach-Object { if($_.Line.split(" ")[1] -eq '3') { Write-Output $_.Line.split(":")[0] } } | Out-File -FilePath $iframelist -Encoding ascii
```

## Application and comparison with `ffprobe`

For the same '[Chameleon](https://we.tl/t-Sr7tLrHR8L)' file, `DGIndexNV` took about _4 seconds_ to produce the index file `Chameleon.dgi`.
Of course this was possible with the help of my Nvidia Graphics card (GTX 1060 6G).

And since the video stream is HEVC, I applied the corresponding script and took almost `0` second to get the same I-frame list file.

```PowerShell
$filename = "Chameleon.dgi"; $iframelist = $filename + " I-frame list.txt"; Select-String -Path $filename "^[0-9]+:" | ForEach-Object { if($_.Line.split(" ")[-2] -eq '2') { Write-Output $_.Line.split(":")[0] } } | Out-File -FilePath $iframelist -Encoding ascii
```

`4 seconds` vs `2 mins 28 seconds`. Can you see the difference? 😅 
Needless to say, this approach is obviously so much more efficient than using `ffprobe`.

## An incomplete working script for StaxRip ##

Although I'd like to include this as a script in StaxRip, there's a couple hurdles _other than the license problem itself_. 😅

1. Get the `dgi` index file name automatically.
2. Run different command lines depending on the source video format.

And so I'd like to see the commented-out holes filled in in the following script:

```PowerShell
$filename = <# %source_file% #>
$iframelist = $filename + " I-frame list.txt"

#
# Check if the source filter is DGSource here.
#

if (<# %source_video_format% is AVC #>) {
    Select-String -Path $filename "^[0-9]+:" | ForEach-Object { if($_.Line.split(" ")[1] -eq '7') { Write-Output $_.Line.split(":")[0] } } | Out-File -FilePath $iframelist -Encoding ascii
} elseif (<# %source_video_format% is HEVC #>) {
    Select-String -Path $filename "^[0-9]+:" | ForEach-Object { if($_.Line.split(" ")[-2] -eq '2') { Write-Output $_.Line.split(":")[0] } } | Out-File -FilePath $iframelist -Encoding ascii
} elseif (<# %source_video_format% is MPG#>) {
    Select-String -Path $filename "^[0-9]+:" | ForEach-Object { if($_.Line.split(" ")[1] -eq 'I') { Write-Output $_.Line.split(":")[0] } } | Out-File -FilePath $iframelist -Encoding ascii
} elseif (<# if video stream is VC1 #>) {
    Select-String -Path $filename "^[0-9]+:" | ForEach-Object { if($_.Line.split(" ")[1] -eq '3') { Write-Output $_.Line.split(":")[0] } } | Out-File -FilePath $iframelist -Encoding ascii
} else {
    [MainModule]::MsgInfo('The video stream type is not supported by DGIndexNV.')
}
```

Thanks for reading this loooooong post. 😅
Your help would be greatly appreciated. 😄

**Reply by ghost** [reply]
So after all ffprobe is not very usable 😞
And sources that do indexing like FFVideoSource seems to have no options for that... Maybe it's too much trouble...
And if the only source that could be made to work is DGSource, then maybe you're going into much trouble for nothing.
If you had your index of I frames, what would you use it for ?
I would guess for cutting on I frames, or previewing I frames only (better performance), right ?

**Reply by JJKylee** [reply]
Yeah. That's exactly the dilemma we're in: `ffprobe` is not practical when it comes to the frame _index_ issues. And the only license-free indexer used for a wide variety of video stream types generating a separate index file, `FFVideoSource`, makes its index file _unreadable_ by encrypting it. So we're left with `DGIndexNV` but this one comes with a license problem. 🤨

> If you had your index of I frames, what would you use it for?

It's useful for _trouble-free stream copy trimming_.

Suppose you know the I-frame index numbers. Then in StaxRip, you can set the _trim_ frame numbers to those I-frames, and go on with _Copy/Mux_ on the video stream to quickly produce a _trouble-free_ video file via `mkvmerge`.

If the frames are set to _non-I_ frames with _Copy/Mux_, you will see the output video either _jumps at the start or starts/ends incorrectly at the play bar_ in many cases. I-frame cutting resolves this issue in stream copy trimming.

**Reply by JJKylee** [reply]
Thanks to [@DJATOM's reminder](https://github.com/staxrip/staxrip/issues/619#issuecomment-806143085), I've also found that we need to consider the situation where a given I-frame is not an IDR-frame (esp. in a HEVC stream) for stream copy cutting.

Now things are getting more complicated. 😅

**Reply by stax76** [reply]
@JJKylee 

I had no time to read the tread because of the task dialog, and today I want to work on keyboard shortcut issues for personal and mpv.net usage. Are there open questions or problems?

**Reply by JJKylee** [reply]
I need your help with these parts:

1. How can I apply the StaxRip macro `%source_file%` in the PowerShell script?
2. How can I check if the `%source_file%` is the type that I want? (`dgi` in my case)
3. How can I determine which type the `%source_video_format%` is among AVC, HEVC, MPG and VC1?

These are all commented out in my experimental PowerShell script code above.

**Comment by stax76** [comment, marked as answer]
Sorry about the delay.

> How can I apply the StaxRip macro %source_file% in the PowerShell script?

If you start an external terminal from staxrip you can access it by the environment variable _source_file_.

If you run the script in the staxrip process using the powershell host of staxrip you can get it like so:

```PowerShell
$activeProject = [ShortcutModule]::p
$activeProject.SourceFile
```

> How can I check if the %source_file% is the type that I want? (dgi in my case)

Check where? In PS like so:

```PowerShell
'aaa.dgi' -match '\.dgi'
```

I don't know how/where you want to run the script.

**Reply by JJKylee** [reply]
Like I said in the post, I'd like to put the script in `%script_dir%` and run it from within StaxRip to make use of some of its features.

As you know, once `DGIndexNV` is set up to run as a preprocessor for some file types, a `dgi` index file is automatically generated as the `"%source_file%"` from within StaxRip. I want to extract frame index numbers from it based on the video stream format(codec).

**Reply by stax76** [reply]
I think you didn't yet describe what exactly your goal is, but it's necessary to know in order to help you.

Script file %script_dir% creates a menu item you can click to start the script, then you want to start a process that will take a while to finish, so cannot wait for the result without freezing the UI thread. When the process ends you have a txt file containing the I-Frames, what do want to do with the I-Frame txt file now?

**Reply by JJKylee** [reply]
> When the process ends you have a txt file containing the I-Frames, what do want to do with the I-Frame txt file now?

I'd like to cut at I-frames given by the text file (choose among the given I-frame indexes) and `Copy/Mux` video streams. I think this is useful for generating a reliable output on closed GOP streams.

**Reply by stax76** [reply]
You want 'Go to next/previous I-Frame' commands/menus/shortcuts in the preview? That can work with two small scripts I think. First script creates the file in the main dialog and second script is integrated in the preview menu. Do you have questions now?

**Reply by JJKylee** [reply]
That'd be great if implemented. 👍👍👍

**Reply by stax76** [reply]
I made 2 changes required for your goal:

1. The script host won't create a new thread but reuses the current thread, this is necessary because UI changes can only be made from the UI thread. I hope this will not break other use cases and scripts.

2. The active form can be assessed in a global variable, so you can call preview commands.

This is all you need I think.

The command to jump to a given frame:

```VB
[ShortcutModule]::g.ActiveForm.SetAbsolutePos(0)
```

The current position:

```VB
[ShortcutModule]::s.LastPosition
```

If you need more help I can help next week or maybe Dendraspis can also help.

https://github.com/staxrip/staxrip/wiki/PowerShell-Scripting

https://www.dropbox.com/sh/4ctl2y928xkak4f/AAADEZj_hFpGQaNOdd3yqcAHa?dl=0

https://1drv.ms/u/s!ArwKS_ZUR01g0kH4d4eT_6a3GaKe?e=qbOfGS

**Reply by JJKylee** [reply]
Thanks. I'll look into it. 😄

**Reply by stax76** [reply]
I made an important improvement, passing arguments in ($args):

https://github.com/staxrip/staxrip/wiki/Commands#executepowershellfile

https://www.dropbox.com/sh/4ctl2y928xkak4f/AAADEZj_hFpGQaNOdd3yqcAHa?dl=0

https://1drv.ms/u/s!ArwKS_ZUR01g0kH4d4eT_6a3GaKe?e=qbOfGS

I sometimes cut with staxrip, so the planed feature is interesting for me too. Normally I would try to help, but I'm currently worried about chest pain, never had this before today. 🙁

**Reply by JJKylee** [reply]
Million thanks. This will make writing/using PowerShell scripts a LOT better. 👍👍👍👍👍

Sorry about the illness. 😥
Hope you get well soon. 😲

**Reply by stax76** [reply]
Thanks I'm fine. Yesterday I was out for a walk, then had a sudden chest pain like a stab, it went away after stopping to walk and came back when I tried to walk again, after a pause I walked home slowly without pain. In the evening it happened again a few time.

I was at the doctor today because of an older appointment and had an ECG check which was OK, she calls me in the evening to tell me the results of the blood test. No pain so far today. I will have to be more careful with my health.

**Reply by JJKylee** [reply]
Good to hear that it was not serious. Hope it was just a temporary upset stomach. 👀

**Reply by ghost** [reply]
I hope you feel better today @stax76 
Try doing only things that please you. It must not be easy to cope with such a chronic condition...
Take good care of yourself 🙏🏼
